### PR TITLE
STAC core part one

### PR DESCRIPTION
**Pull request by uclaros**
## Description
This is the first part of the implementation of https://github.com/qgis/QGIS-Enhancement-Proposals/issues/300
It includes the building block classes for parsing and handling STAC elements and allows interacting with STAC catalogs using the QGIS browser.

The contents of both static catalogs and STAC API endpoints can be retrieved using the new STAC root entry on the browser.
image

STAC Items that contain assets that are cloud optimized (COG, COPC, EPT for now) can be drag and dropped to the map canvas and those assets will be added as layers.
Peek 2024-10-04 15-51

STAC Item assets may be downloaded to a local folder using the _Download Assets_ dialog.
image
image

All Catalogs, Collections and Items properties can be examined using the STAC Object Details dialog:
Peek 2024-10-04 15-56

In subsequent PRs the Data Source Manager implementation will be added which will also allow searching for Items and Collections

Any feedback is warmly welcome!

### CODE REVIEW

**Review by troopa81 — COMMENTED** [review]
Really happy to see this landing in core, thanks for this work! 

> Local files reading and blocking network requests in QgsStacController

Could we avoid blocking request and use QgsTask ? I can see no way of canceling (I made a very very quick read of this PR though, I made have skipped it), what happen if I want to close the stac catalog window but QGIS is currently retrieving a lot of items?

**Comment by uclaros** [comment]
> Could we avoid blocking request and use QgsTask ? I can see no way of canceling (I made a very very quick read of this PR though, I made have skipped it), what happen if I want to close the stac catalog window but QGIS is currently retrieving a lot of items?

Blocking requests will only be used for the initial implementation on the browser panel, where `QgsDataItem::createChildrent()` is always called on a separate thread and there is no concept of closing/canceling in the browser.
The data source manager implementation, which will also support searching, will use async requests which could potentially be cancelled when the data source manager is closed.

**Comment by saberraz** [comment]
Nice work! 
Just a minor comment on UI/UX: in the Browser panel, at the moment, :information_source:  icon is shown for the datasets:
image

Some datasets do have a preview image. It would be good to use the preview image instead of the :information_source: icon:
image

**Comment by uclaros** [comment]
> Some datasets do have a preview image.

The STAC spec suggests to use `thumbnail` as a role for a quick preview image of the data which should be less than 600x600 pixels. Assets with this role are rendered on top of the Object Details dialog when available.
The `overview` role may contain large images and is probably more appropriate to actually load as a layer.

In any case, even the `thumbnail` images may be too big to be used as icons for the DataItems. For me it would make more sense to try and display those `thumbnails` on the item's tooltips instead.

File: src/core/stac/qgsstaccontroller.h [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+    /**
+     * Sets the authentication config id which should be used during the request.
+     * \see authCfg()
+     */
+    void setAuthCfg( const QString &authCfg );
+
+  signals:
+    void finishedStacObjectRequest( int id, QString errorMessage );
+    void finishedItemCollectionRequest( int id, QString errorMessage );
```
Comment (wonder-sk) [inline src/core/stac/qgsstaccontroller.h:128]: worth commenting these two signals...

File: src/core/stac/qgsstacobject.h [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    //! Return the url stored in the object's "root" link
+    QString rootUrl() const;
+
+    //! Return the url stored in the object's "parent" link
+    QString parentUrl() const;
+
+
+  protected:
+    bool mValid = false;
+    Type mType = Type::Unknown;
```
Comment (wonder-sk) [inline src/core/stac/qgsstacobject.h:101]: Shall we make a rule that "unknown type" == not valid? we could then get rid of the mValid member...

**Review by wonder-sk — APPROVED** [review]
This looks good to me now

(we have discussed the work with Stefanos extensively, so my earlier comments have been addressed already)

If nobody objects, I would merge this tomorrow...

**Pull request merged**